src/mn_menu.c

```diff
@@ -517,6 +517,32 @@ void M_ClearEpisodes(void)
     NewDef.prevMenu = &MainDef;
 }
 
+int MN_GetEpisodeMap(int map)
```

Comment:
What exactly did this code do? Could you add a comment, please?

Reply:
The problem is that id24 spec resets the map numbers in the episodes. For example, if we have an episode 2 definition on MAP15, then the intermission screen spec considers MAP16 as MAP01 of current episode.

I added comment here: https://github.com/fabiangreffrath/woof/pull/1833/commits/896fe99a3b3b9c6be0316116dee534db1cd3334e

Reply:
Alright, thank you. 

But if we are in e.g. MAP15 and there is a new episode start at both MAP10 and MAP16, it will return MAP10, right?

Reply:
> But if we are in e.g. MAP15 and there is a new episode start at both MAP10 and MAP16, it will return MAP10, right?

Yes, at MAP15 we are still in MAP10 episode. 

It is probably possible to write UMAPINFO where this approach would not work properly.

Reply:
> It is probably possible to write UMAPINFO where this approach would not work properly.

Pretty sure. For example, if you construct a secret route across higher level numbers (e.g. the IWAD's 15->31->32->16 route).

Reply:
A solution would be to go through all the `next` and `nextsecret` fields in UMAPINFO  (also considering default progression) and build arrays of maps for all episodes.

Reply:
You could iterate through all the maps that have an episode start, from each of these maps iterate through the secret route (`nextsecret`) and through the regular route (`G_GotoNextLevel(int *pEpi, int *pMap)` will come in handy) and if the map in question comes up, return the current episode.

Reply:
Next question, what if a map takes part in two routes?

Reply:
> Next question, what if a map takes part in two routes?

I guess we need to remeber the selected episode and keep it in save.

By the way, `-warp` doesn't work properly in this case. Perhaps we don't need to worry about it.

Reply:
But if you finish a map that you have warped to and you reach the intermission screen, it should know the right episode.

Reply:
> But if you finish a map that you have warped to and you reach the intermission screen, it should know the right episode.

How, if this map is part of two episodes? The syntax `-warp E M` works only for Doom 1.

Reply:
Oh sure, I was talking about the unambiguous case. In the ambiguous case we just need to decide e.g. if we want to return the highest or lowest possible episode number.

Reply:
However, this approach will be vulnerable against loops in the map progression. For example, if you have to revisit the same map again and again until you find the secret exit.

Reply:
I don't think it's possible to solve all cases. For example, the UMAPINFO update for the classic [Skillsaw WADs](https://www.doomworld.com/profile/8877-skillsaw/?status=1747176&type=status) doesn't have an episode ending definitions, so there's no way to know where the episode ends. These WADs have death exits, so they don't need episode ending.

Perhaps the current `MN_GetEpisodeMap` is okay for most real-world cases.

Reply:
Yes okay. Let's not try to solve problems before they exist.

Reply:
Check out [Fork in the Road](https://www.doomworld.com/forum/topic/107191-fork-in-the-road-a-umapinfo-project-2019-07-03-update-alpha-3/). How should the `G_NextLevel` function work in this case?

Reply:
> Yes okay. Let's not try to solve problems before they exist.

I have an idea. We will only consider `player->visitedlevels`, which will be reset every time a map with `episode` or `endgame` definition is encountered. As for `-warp` or IDCLEV, we will simply not show id24 intermission in case of Doom 2.

Reply:
Hm, I guess the whole concept of "map number in episode = map number - number of first map in episode" would break if a map transition would cross episode boundaries. That's probably not going to work with ID24 anyway. :shrug:

Reply:
From [ID24 spec](https://docs.google.com/document/d/1IVJ4nt5f20_NUrMrr9gl-UDStq1jexo5y4XTL1qd4vs/edit?usp=sharing):
> The map number accessible for condition checking is dependent on whether the victory
screen is currently displaying an exit animation or an enter animation. If it is an exit
animation, the map number corresponds to the map being played; if it is an enter animation,
the map number corresponds to the map about to be entered. The map number itself is
defined through other means, including but not limited to UMAPINFO, and is outside of the
scope of this document to define.

From my expiriments with the KEX port, it seems that the map number is determined by the order of the map definitions in the UMAPINFO lump, the episode ends when another `episode` field is encountered. The `next` and `nextsecret` fields are ignored when determining map numbers.

So my suggestion is to define map numbers from actual numbers in UMAPINFO. We can detect "Legacy of Rust" and make a compatibility hack for it.

Reply:
@fabiangreffrath Here I have emulated the behavior of the KEX port as far as I could understand it: https://github.com/fabiangreffrath/woof/pull/1833/commits/e87cc4c8bd04c1487f42c26e6f5d3fa6de865f6e

Reply:
Impressive, thank you!